### PR TITLE
Weights & Biases Basic Integration

### PR DESCRIPTION
**Pull request by kshen3778**
# Pull Request Template

## Description

This PR adds a Weights & Biases logger (WandbLogger) and integrates it into KerasModel and TorchModel

After discussion on the developer call, it was suggested that a new class be created for W&B logging, which we have created in `wandblogger.py` and contains all the functionalities for interacting with W&B (sending information to a user's runs and dashboard).

Currently, it supports basic W&B logging. Future upgrades will allow for W&B Artifacts integration (model checkpointing, dataset logging, versioning)

Changes made:
- Added WandbLogger (logs training loss during training and any values calculated in ValidationCallback)
- Modified KerasModel and TorchModel to check wandb installation and setup WandbLogger
- Unit tests
- Modified .rst docs + docstrings
- Previous wandb flag argument has been kept but marked as deprecated

Additional local tests for functionality (Colab): https://colab.research.google.com/drive/1fpS87sGsCsoWP-4F48SVgmLRrmC_bT3x?usp=sharing

## Type of change

Please check the option that is related to your PR.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
    - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ]  Documentations (modification for documents)

## Checklist

- [x]  My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
    - [x]  Run `yapf -i <modified file>` and check no errors (**yapf version must be 0.22.0**)
    - [x]  Run `mypy -p deepchem` and check no errors
    - [x]  Run `flake8 <modified file> --count` and check no errors
- [x]  I have performed a self-review of my own code
- [x]  I have commented my code, particularly in hard-to-understand areas
- [x]  I have made corresponding changes to the documentation
- [x]  I have added tests that prove my fix is effective or that my feature works
- [x]  New unit tests pass locally with my changes
- [x]  I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by rbharath** [comment]
Great to see this up! Let me know when this is ready for us to do a review pass.

**Comment by kshen3778** [comment]
Hey @rbharath! I just added some tests today so it should be good for a review. We decided to make our changes into a draft PR first since we plan on adding more things later on, but we want to make sure the changes made so far are acceptable. Let us know if this makes sense or if you prefer us to just continue.

File: deepchem/models/keras_model.py [hunk 1]
```diff
@@ -131,7 +120,7 @@ def __init__(self,
                learning_rate: Union[float, LearningRateSchedule] = 0.001,
                optimizer: Optional[Optimizer] = None,
                tensorboard: bool = False,
-               wandb: bool = False,
+               wandb_logger: Optional[WandbLogger] = None,
```
Comment (rbharath) [inline deepchem/models/keras_model.py:123]: For backwards compatibility, we would need to continue supporting the previous flag as well. You can add a deprecation warning on it (we can only make breaking changes at the major release version so can't remove it till DeepChem 3)

Comment (rbharath) [reply deepchem/models/keras_model.py:123]: One thought is that if `wandb` is set, we instantiate a default logger here

**Review by rbharath — COMMENTED** [review]
Did a quick pass and so just made a comment about backwards compatibility. Can do a more thorough review once this is ready 

CC @peastman who might be interested as well

File: deepchem/models/keras_model.py (hunk 1 above)
Comment (kshen3778) [reply deepchem/models/keras_model.py]: Sure, I've re-add support for the wandb flag back into fit() and ValidationCallback along with deprecation warning.

**Comment by peastman** [comment]
This looks like a good start!  My one major comment is that this API seems unnecessarily Wandb-specific.  We currently try to provide a uniform API that treats all logging mechanisms equivalently.  For example, if you want to perform validation, you add a ValidationCallback.  That callback can log information to whatever target you requested: a file, the console, Wandb, or TensorBoard.  You've created a new validation and logging API, such that logging to Wandb is done with completely different classes than logging to TensorBoard.

I would prefer to keep the uniform API.  More specifically, the following should be specified independently:

1. *What* to log (validation set, metrics, etc.)
2. *Where* to log it (Wandb, TensorBoard, etc.)

Either or both of those can be extended from what we have right now, which is pretty limited.  But I'd keep them separate from each other.  WandbLogger would specify only information that's specific to Wandb.  The choice of what information to log would continue to be specified with a uniform API, independent of where you're logging it to.

**Comment by kshen3778** [comment]
@peastman I looked over the code to see if we could separate out the functionalities, but a few challenges popped into my mind as I was refactoring. It mainly concerns the limited functionality of the current uniform logging API (especially for logging training data). 

The reason we moved a lot of functionality into the WandbLogger is because we wanted access to a lot of things that are not provided by KerasModel. For example, having access to the datasets and metrics in order to run evaluate() and log metrics for the training loop. We also add some functionalities such as calculating the current epoch and how many samples the model has seen. We don't have access to datasets and metrics inside KerasModel's fit_generator() so that's why we passed them into WandbLogger upon initialization, so we would have access to these items.

Another feature we would want the WandbLogger to have is the option to specify the logging strategy between "steps" and "epoch". We find that a lot of ML developers log by epoch and not steps. Currently, KerasModel's fit_generator() only logs by steps.

If we were to stick to the uniform API, I'm afraid we would be limited to just logging the batch loss during training, unless you are fine with us modifying KerasModel fit() and fit_generator(). Basically, the current API used for training logging is quite limited, and we are forced to calculate a lot of the items inside our WandbLogger (epoch_number, scores). If we had access to the "datasets" and "metrics" inside fit_generator() it would've been much easier.

Validation is a different story and we might be able to integrate WandbLogger into the ValidationCallback, since inside ValidationCallback we have access to "datasets" and "metrics".

Perhaps it would be best if we kept the current WandbLogger implementation for training, and then I will see if I can move validation functionality into ValidationCallback?

**Comment by peastman** [comment]
`fit_generator()` doesn't know anything about Datasets, because there might not be one.  The generator can be anything.  It might be a piece of code that programmatically generates infinite training data.  In that case, epochs are also undefined.  Every sample is unique and it never repeats.

There's definitely room to improve the current logging support, which is kind of limited.  We just need to do it in a way that isn't unnecessarily tied to a specific logging mechanism.  For example, you suggested an option to log every N epochs rather than every N steps.  That's a great idea.  If you want to implement it, that would be a nice improvement.  But it applies just as well to console output or TensorBoard as to Wandb, so it should be specified with a generic API and work the same way whatever logging mechanism you use.

**Comment by kshen3778** [comment]
> `fit_generator()` doesn't know anything about Datasets, because there might not be one. The generator can be anything. It might be a piece of code that programmatically generates infinite training data. In that case, epochs are also undefined. Every sample is unique and it never repeats.
> 
> There's definitely room to improve the current logging support, which is kind of limited. We just need to do it in a way that isn't unnecessarily tied to a specific logging mechanism. For example, you suggested an option to log every N epochs rather than every N steps. That's a great idea. If you want to implement it, that would be a nice improvement. But it applies just as well to console output or TensorBoard as to Wandb, so it should be specified with a generic API and work the same way whatever logging mechanism you use.

Sounds good, we'll give it a look and make the necessary changes.

**Comment by kshen3778** [comment]
Hey @peastman, we did some refactoring and made the necessary changes. Feel free to take a look and let us know if it's acceptable (we did have to go into KerasModel `fit()` and add a few lines in order to open up access to the dataset in order to log epoch-wise and specify metrics to log during training). If all looks fine, I can at the earliest time possible make the same integration into TorchModel, then mark the PR as ready for review so it can be merged.

Some other questions:
1. Regarding the documentation, are there any other places we would need to modify apart from docstrings? I know in the contributing guide it says the docs are built automatically from docstrings, but I also see a lot of .rst files [here](https://github.com/deepchem/deepchem/tree/master/docs/source/api_reference) so I'm wondering if there are more places we would need to add documentation for our changes.

2. On the PR it says I would need to run 2 workflows and a maintainer must first approve the workflows for first-time contributors. Would it be possible to get access to those workflows?

Thanks!

**Comment by peastman** [comment]
Thanks, this is looking good!  More specific comments to follow.

File: deepchem/models/callbacks.py [hunk 2]
```diff
@@ -49,6 +48,10 @@ def __init__(self,
       if True, the best model is considered to be the one that minimizes the
       validation metric.  If False, the best model is considered to be the one
       that maximizes it.
+    logging_strategy: str
+      the logging strategy used for logging (step or epoch). If "step",
+      logging interval will be the value provided for `interval`. If "epoch",
+      then logging will happen at the end of every training epoch.
```
Comment (peastman) [inline deepchem/models/callbacks.py:54]: I'm not sure this is exactly the behavior we want.  I suggest this flag determines whether `interval` is a number of steps or a number of epochs.  When training with a small dataset, you might only want to perform validation every 10 or 100 epochs.

Another option would be to deprecate the `interval` argument and replace it with two different arguments: `step_interval` and `epoch_interval`.  The user would specify one or the other, but not both.

File: deepchem/models/keras_model.py [hunk 3]
```diff
@@ -133,6 +136,8 @@ def __init__(self,
                tensorboard: bool = False,
                wandb: bool = False,
                log_frequency: int = 100,
+               logging_strategy: Optional[str] = "step",
```
Comment (peastman) [inline deepchem/models/keras_model.py:139]: This argument needs to be documented.  The documentation for `log_frequency` also needs to change.

File: deepchem/models/keras_model.py [hunk 4]
```diff
@@ -312,16 +351,22 @@ def fit(self,
...
+    metrics: Optional[List[Metric]], optional (default None)
+      metrics to compute on the dataset used during training. If None,
+      no metrics and scores will be computed and only training loss will be logged.
 
     Returns
     -------
     The average loss over the most recent checkpoint interval
    """
+    self.dataset = dataset
+    self.metrics = metrics
```
Comment (peastman) [inline deepchem/models/keras_model.py:363]: This isn't going to work.  We can't assume the user calls `fit()` and passes in a dataset and metrics.  They might call `fit_generator()` directly.  In that case, there is no dataset and epochs are undefined.

File: deepchem/models/keras_model.py [hunk 5]
```diff
@@ -426,20 +474,50 @@ def fit_generator(self,
...
+      # Calculate epoch number, sample count, and metrics
+      epoch_num = self._get_epoch_num(current_step)
+      sample_count = self._get_sample_count(current_step)
+
+      # Decide whether to calculate metrics at this current step
+      scores = None
+      if self.metrics is not None and self.metrics:
+        if (self.logging_strategy == "step" and current_step % self.log_frequency == 0) or \
+           (self.logging_strategy == "epoch" and current_step % steps_per_epoch == 0):
+          scores = self.evaluate(self.dataset, self.metrics)
```
Comment (peastman) [inline deepchem/models/keras_model.py:486]: I don't think this is the right way or place to be computing metrics.  If you're computing metrics, you don't necessarily want to compute them on the training set (which again might not even be a Dataset object), and you might not want to compute them at the same interval at which you record training loss (since this is much more expensive).

**Comment by peastman** [comment]
After looking through the code some more, I suggest we separate out all the changes to logging behavior, like logging per epoch rather than per step.  It's adding a lot of complexity and we need to think some more about the design.  And it isn't really specific to Wandb.  It will be simpler if we do a first PR that is just Wandb-specific changes while leaving the basic logging behavior the same, and then a separate PR with improvements to the behavior.

I notice a lot of formatting changes.  We run all the code through yapf 0.22.  If you use a different version, CI will complain about it.  (Though we really should upgrade to a newer version one of these days.)

File: deepchem/models/wandblogger.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def is_wandb_available():
+  return importlib.util.find_spec("wandb") is not None
+
+
+class WandbLogger(object):
+  """Weights & Biases Logger for KerasModel.
```
Comment (peastman) [inline deepchem/models/wandblogger.py:14]: It should work with all models, not just KerasModel.  We also have a TorchModel class, and we're in the process of creating a JAXModel class.

File: deepchem/models/keras_model.py (hunk 4 above)
Comment (kshen3778) [reply deepchem/models/keras_model.py:363]: Is there a preferred way for us to calculate the epoch number? We would need this value for logging based on epoch interval. Currently, we are calculating it based on the size of the training dataset. But as you mentioned, the training dataset may not exist if they are calling `fit_generator()` directly and not calling `fit()`

It's not clear to me whether there is a good way to calculate the epoch number directly inside `fit_generator()`. We could just check to see if the user has started training via `fit()` and in that case they would have passed in a Dataset object.

If there isn't a clear way, we can just remove the epoch logging option entirely and go back to only supporting step intervals.

File: deepchem/models/keras_model.py (hunk 5 above)
Comment (kshen3778) [reply deepchem/models/keras_model.py:486]: In your opinion, what do you think is the best approach to log metrics for training? I know that some users prefer to see metrics being calculated on the training dataset during training too. From what I understand, there are no metrics that are passed into `fit()` originally so there would need to be a place to pass in what metrics to log.

File: deepchem/models/callbacks.py (hunk 2 above)
Comment (kshen3778) [reply deepchem/models/callbacks.py:54]: > After looking through the code some more, I suggest we separate out all the changes to logging behavior, like logging per epoch rather than per step. It's adding a lot of complexity and we need to think some more about the design. And it isn't really specific to Wandb. It will be simpler if we do a first PR that is just Wandb-specific changes while leaving the basic logging behavior the same, and then a separate PR with improvements to the behavior.
> 
> I notice a lot of formatting changes. We run all the code through yapf 0.22. If you use a different version, CI will complain about it. (Though we really should upgrade to a newer version one of these days.)

Ah my bad, I just saw this comment. Sure, we can remove all the changes for logging per epoch and also logging training metrics. I agree that it would be more complex and require more extensive modifications. We could do a simple PR first with just the current logging behaviour and go from there.

**Comment by kshen3778** [comment]
> After looking through the code some more, I suggest we separate out all the changes to logging behavior, like logging per epoch rather than per step. It's adding a lot of complexity and we need to think some more about the design. And it isn't really specific to Wandb. It will be simpler if we do a first PR that is just Wandb-specific changes while leaving the basic logging behavior the same, and then a separate PR with improvements to the behavior.
> 
> I notice a lot of formatting changes. We run all the code through yapf 0.22. If you use a different version, CI will complain about it. (Though we really should upgrade to a newer version one of these days.)

Ah my bad, I just saw this comment. Please feel free to ignore my previous replies above. Sure, we can remove all the changes for logging per epoch and also logging training metrics. I agree that it would be more complex and require more extensive modifications. We could do a simple PR first with just the current logging behaviour and go from there.

File: deepchem/models/wandblogger.py (hunk 6 above)
Comment (kshen3778) [reply deepchem/models/wandblogger.py:14]: Will do! We'll also add the integration to TorchModel shortly.

**Comment by kshen3778** [comment]
@peastman Made the changes. I've stripped out all the logging related modifications you mentioned earlier. I've also re-ran with yapf 0.22 so the styling should be fixed.

It's ready for review. Feel free to give feedback

**Comment by peastman** [comment]
This is looking really good!  I think we're getting close.  Just a few minor suggestions.

Instead of handling the `wandb` and `wandb_logger` arguments separately, it might be simplest to just create a default logger object if they specified the deprecated argument?  Then we wouldn't need separate code for the two arguments in every place that does logging.

You deleted some blank lines from comments in callbacks.py.  Could you put them back?

In the documentation, it's probably ok to only describe the new method.  We don't need to include the deprecated one as well.

**Comment by peastman** [comment]
`test_wandblogger()` is failing on CI because it isn't installed.  You have the check

```python
@unittest.skipIf(not has_pytorch, 'PyTorch is not installed')
```

It should be checking for wandb, not pytorch.

**Comment by kshen3778** [comment]
> This is looking really good! I think we're getting close. Just a few minor suggestions.
> 
> Instead of handling the `wandb` and `wandb_logger` arguments separately, it might be simplest to just create a default logger object if they specified the deprecated argument? Then we wouldn't need separate code for the two arguments in every place that does logging.
> 
> You deleted some blank lines from comments in callbacks.py. Could you put them back?
> 
> In the documentation, it's probably ok to only describe the new method. We don't need to include the deprecated one as well.

Yup sure, I've went ahead and removed the deprecated code entirely in order to use a default logger. Let me run a few local tests and get back to you - I've just pulled some commits from masters so I just want to make sure everything still works.

**Comment by kshen3778** [comment]
> `test_wandblogger()` is failing on CI because it isn't installed. You have the check
> 
> ```python
> @unittest.skipIf(not has_pytorch, 'PyTorch is not installed')
> ```
> 
> It should be checking for wandb, not pytorch.

Just to check, would something like 
`@unittest.skipIf((not has_pytorch) or (not has_wandb), 'PyTorch and/or Wandb is not installed')`
make more sense since we are checking for both wandb and Pytorch (since all the other tests check for Pytorch too)

**Comment by rbharath** [comment]
@kshen3778 Looks like that test uses pytorch so you'll need both guardrails I think

File: deepchem/models/keras_model.py [hunk 7]
```diff
@@ -181,14 +186,41 @@ def __init__(self,
       self.optimizer = optimizer
     self.tensorboard = tensorboard
 
-    # W&B logging
+    # W&B flag support (DEPRECATED)
+    if wandb:
+      logger.warning(
+          "`wandb` argument is deprecated. Please use `wandb_logger` instead. "
+          "This argument will be removed in a future release of DeepChem.")
```
Comment (peastman) [inline deepchem/models/keras_model.py:193]: TorchModel should also display this warning.

Comment (kshen3778) [reply deepchem/models/keras_model.py:193]: Added! Must've missed it earlier

**Comment by kshen3778** [comment]
> @kshen3778 Looks like that test uses pytorch so you'll need both guardrails I think

Sure, I've added in both guardrails.

**Comment by peastman** [comment]
`test_wandblogger()` in test_kerasmodel.py failed.  It needs a guard too.

**Comment by kshen3778** [comment]
> `test_wandblogger()` in test_kerasmodel.py failed. It needs a guard too.
I've went ahead and added wandb guardrail into test_kerasmodel

**Comment by peastman** [comment]
This looks good, and tests are passing except for the ones we know are failing for other reasons.  It's ready to merge as far as I'm concerned.

**Comment by kshen3778** [comment]
Perfect thanks, feel free to merge it when ready. I don't expect any changes on my end in the near future.

**Comment by rbharath** [comment]
Ok merged! This will be in our next release (which is unfortunately delayed due to TensorFlow issues https://forum.deepchem.io/t/deepchem-2-6-0-delayed-due-to-tensorflow-numpy-1-20/503). We have some mitigations and fixes on the way so hopefully we'll be back to our release schedule within a couple of months
